crates/shared/src/event_handling.rs

```diff
@@ -152,14 +152,12 @@ where
         }
 
         // handle special case when no new block is added
+        // this case would be caught later in algorithm by `detect_reorg_path`,
+        // but we skip some node calls by returning early
         if (current_block_number, current_block_hash)
             == (last_handled_block_number, last_handled_block_hash)
         {
-            return Ok(EventRange {
-                history_range: None,
-                latest_blocks: vec![],
-                is_reorg: false,
-            });
+            return Err(anyhow!("update entered with already handled block"));
```

Comment:
Why do we need to return an error here? Isn't updating to the current block a no-op, which should be OK?

The reason I ask is that:
1. `ServiceMaintaining` sees block N
    1. Event updater A runs and succeeds
    2. Event updater B runs, but gets request sent to node that is on block N-1 and fails
2. `ServiceMaintaining` retries because of event updater B failure
    1. Event updater A runs, but its already updated and returns an error

Reply:
Ah yes, your case is problematic because multiple EventHandlers share common db.

I was aiming to fix the problem when multiple EventHandlers with no common storage (Balancer for example having EventHandler for each factory) enter the update for block N, and some of them detects the current_block  as N-1, then it would return an error and retry only that one.

I need to think how to cover both cases. Maybe the solution is to send the block number for which we are updating to the EventHandler so we can differentiate between:
1. entered update for block N, last handled is N, current_block in EventHandler says N - this is Ok
2. entered update for block N+1, last handled is N, current_block in EventHandler says N - this is error

Reply:
I'm concerned that this is getting too complicated 😞. Personally, I would re-structure this such that:
- We **only** write to database on success, and make sure nothing is written in case of failure
- We only update the `last_seen_blocks` on a successful write to the database and then return right away.

This way,
- the `last_seen_blocks` is always an accurate representation of what was in the database
- the method becomes idempotent (i.e. updating the same block multiple times always yields the same result).

This makes it much easier to reason about failure cases if you ensure that we have "atomic application of updates" internally to the event handler - i.e. it is not left in some undefined, half-updated state because of an internal failure.

Reply:
Specifically, I also wouldn't treat block propagation as a special kind of error. They should be handled (IMO) in the same way as "intermittent network error" - they happen, but things should continue to function when they do.

Reply:
Hmm, maybe I didn't understand the change well.

I still think we have the issue right now that if we fail on a `ServiceMaintenance` run, that we will only retry on the next block, and keep stale data around for the full block. Can this PR be re-purposed to retry maintenance in case it fails? I feel like (if we have atomic updates of event data + last seen block) then this should cover the case completely right? Or am I missing something?